cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -99,82 +111,145 @@ protected void startUp() throws Exception {
       .build();
 
     httpService.startAndWait();
-    running = true;
+
+    executor = Executors.newSingleThreadExecutor();
   }
 
   @Override
   protected void shutDown() throws Exception {
     httpService.stopAndWait();
+    shutdownExecutor();
+  }
+
+  private void executeWorkflowAction(ApplicationSpecification appSpec, WorkflowActionNode node,
+                                     InstantiatorFactory instantiator, ClassLoader classLoader) throws Exception {
+
+    WorkflowActionSpecification actionSpec;
+    ScheduleProgramInfo actionInfo = node.getProgram();
+    switch (actionInfo.getProgramType()) {
+      case MAPREDUCE:
+        MapReduceSpecification mapReduceSpec = appSpec.getMapReduce().get(actionInfo.getProgramName());
+        String mapReduce = mapReduceSpec.getName();
+        actionSpec = new DefaultWorkflowActionSpecification(new ProgramWorkflowAction(
+          mapReduce, mapReduce, SchedulableProgramType.MAPREDUCE));
+        break;
+      case SPARK:
+        SparkSpecification sparkSpec = appSpec.getSpark().get(actionInfo.getProgramName());
+        String spark = sparkSpec.getName();
+        actionSpec = new DefaultWorkflowActionSpecification(new ProgramWorkflowAction(
+          spark, spark, SchedulableProgramType.SPARK));
+        break;
+      case CUSTOM_ACTION:
+        actionSpec = node.getActionSpecification();
+        break;
+      default:
+        LOG.error("Unknown Program Type '{}', Program '{}' in the Workflow.", actionInfo.getProgramType(),
+                  actionInfo.getProgramName());
+        throw new IllegalStateException("Workflow stopped without executing all tasks");
+    }
+
+    WorkflowStatus workflowStatus = new WorkflowStatus(state(), actionSpec, node.getNodeId());
+    status.put(node.getNodeId(), workflowStatus);
+
+    WorkflowAction action = initialize(actionSpec, classLoader, instantiator);
+    try {
+      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(action.getClass().getClassLoader());
+      try {
+        action.run();
+      } finally {
+        ClassLoaders.setContextClassLoader(oldClassLoader);
+      }
+    } catch (Throwable t) {
+      LOG.warn("Exception on WorkflowAction.run(), aborting Workflow. {}", actionSpec);
+      // this will always rethrow
+      throw Throwables.propagate(t.getCause());
+    } finally {
+      // Destroy the action.
+      destroy(actionSpec, action);
+    }
+    status.remove(node.getNodeId());
```

Comment:
If the action throws exception, then it's not removed from the status map. Is it intentional?